# 01-Intro - Homework

In [1]:
# import required libraries
import numpy as np
import pandas as pd

## Q1. Pandas version
What's the version of Pandas that you installed?

You can get the version information using the `__version__` field:

In [2]:
pd.__version__

'2.2.2'

## Getting the data
For this homework, we'll use the Laptops Price dataset. Download it from [here](https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv).

In [3]:
!wget -P ../../datasets/ "https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv"

--2024-09-30 19:45:21--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 298573 (292K) [text/plain]
Saving to: ‘../../datasets/laptops.csv’

laptops.csv         100%[===================>] 291.58K  --.-KB/s    in 0.004s  

2024-09-30 19:45:22 (73.5 MB/s) - ‘../../datasets/laptops.csv’ saved [298573/298573]



Now read it with Pandas.

In [4]:
laptops_df = pd.read_csv("../../datasets/laptops.csv")
laptops_df

Laptop       Status   Brand  \
0     ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...          New    Asus   
1     Alurin Go Start Intel Celeron N4020/8GB/256GB ...          New  Alurin   
2     ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...          New    Asus   
3     MSI Katana GF66 12UC-082XES Intel Core i7-1270...          New     MSI   
4     HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...          New      HP   
...                                                 ...          ...     ...   
2155  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2156  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2157  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2158  Razer Book 13 Intel Evo Core i7-1165G7/16GB/1T...  Refurbished   Razer   
2159  Razer Book FHD+ Intel Evo Core i7-1165G7/16GB/...  Refurbished   Razer   

           Model                CPU  RAM  Storage Storage type       GPU  \
0     ExpertBook      Intel Core i5    8      512          SSD       NaN   
1             Go      Intel Celeron    8      256          SSD       NaN   
2     ExpertBook      Intel Core i3    8      256          SSD       NaN   
3         Katana      Intel Core i7   16     1000          SSD  RTX 3050   
4            15S      Intel Core i5   16      512          SSD       NaN   
...          ...                ...  ...      ...          ...       ...   
2155       Blade      Intel Core i7   16     1000          SSD  RTX 3060   
2156       Blade      Intel Core i7   16     1000          SSD  RTX 3070   
2157       Blade      Intel Core i7   32     1000          SSD  RTX 3080   
2158        Book  Intel Evo Core i7   16     1000          SSD       NaN   
2159        Book  Intel Evo Core i7   16      256          SSD       NaN   

      Screen Touch  Final Price  
0       15.6    No      1009.00  
1       15.6    No       299.00  
2       15.6    No       789.00  
3       15.6    No      1199.00  
4       15.6    No       669.01  
...      ...   ...          ...  
2155    17.3    No      2699.99  
2156    17.3    No      2899.99  
2157    17.3    No      3399.99  
2158    13.4   Yes      1899.99  
2159    13.4   Yes      1699.99  

[2160 rows x 12 columns]

We can also read directly from the [link](https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv) using the following command:
```
url = "https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv"

laptops_df = pd.read_csv(url)

```

Display top 5 rows from the Laptops DataFrame.

In [5]:
laptops_df.head()

Laptop Status   Brand  \
0  ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...    New    Asus   
1  Alurin Go Start Intel Celeron N4020/8GB/256GB ...    New  Alurin   
2  ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...    New    Asus   
3  MSI Katana GF66 12UC-082XES Intel Core i7-1270...    New     MSI   
4  HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...    New      HP   

        Model            CPU  RAM  Storage Storage type       GPU  Screen  \
0  ExpertBook  Intel Core i5    8      512          SSD       NaN    15.6   
1          Go  Intel Celeron    8      256          SSD       NaN    15.6   
2  ExpertBook  Intel Core i3    8      256          SSD       NaN    15.6   
3      Katana  Intel Core i7   16     1000          SSD  RTX 3050    15.6   
4         15S  Intel Core i5   16      512          SSD       NaN    15.6   

  Touch  Final Price  
0    No      1009.00  
1    No       299.00  
2    No       789.00  
3    No      1199.00  
4    No       669.01

## Q2. Records count
How many records are in the dataset?

- 12
- 1000
- 2160
- 12160

In [6]:
len(laptops_df)

2160

## Q3. Laptop brands
How many laptop brands are presented in the dataset?

- 12
- 27
- 28
- 2160

In [7]:
laptops_df['Brand'].nunique()

27

In [8]:
laptops_df['Brand'].unique()

array(['Asus', 'Alurin', 'MSI', 'HP', 'Lenovo', 'Medion', 'Acer', 'Apple',
       'Razer', 'Gigabyte', 'Dell', 'LG', 'Samsung', 'PcCom', 'Microsoft',
       'Primux', 'Prixton', 'Dynabook Toshiba', 'Thomson', 'Denver',
       'Deep Gaming', 'Vant', 'Innjoo', 'Jetwing', 'Millenium', 'Realme',
       'Toshiba'], dtype=object)

## Q4. Missing values
How many columns in the dataset have missing values?

- 0
- 1
- 2
- 3

In [9]:
laptops_df.columns

Index(['Laptop', 'Status', 'Brand', 'Model', 'CPU', 'RAM', 'Storage',
       'Storage type', 'GPU', 'Screen', 'Touch', 'Final Price'],
      dtype='object')

In [10]:
laptops_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2160 entries, 0 to 2159
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Laptop        2160 non-null   object 
 1   Status        2160 non-null   object 
 2   Brand         2160 non-null   object 
 3   Model         2160 non-null   object 
 4   CPU           2160 non-null   object 
 5   RAM           2160 non-null   int64  
 6   Storage       2160 non-null   int64  
 7   Storage type  2118 non-null   object 
 8   GPU           789 non-null    object 
 9   Screen        2156 non-null   float64
 10  Touch         2160 non-null   object 
 11  Final Price   2160 non-null   float64
dtypes: float64(2), int64(2), object(8)
memory usage: 202.6+ KB


In [11]:
# We can also check the missing value count using the following command:
laptops_df[laptops_df.columns].isnull().sum()

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

As we can see there are 3 columns with the missing values: Storage type, GPU and Screen.

## Q5. Maximum final price
What's the maximum final price of Dell notebooks in the dataset?

- 869
- 3691
- 3849
- 3936

In [12]:
laptops_df[laptops_df['Brand']=='Dell']['Final Price'].max()

3936.0

## Q6. Median value of Screen
1. Find the median value of `Screen` column in the dataset.
2. Next, calculate the most frequent value of the same `Screen` column.
3. Use `fillna` method to fill the missing values in `Screen` column with the most frequent value from the previous step.
4. Now, calculate the median value of `Screen` once again.
Has it changed?

> Hint: refer to existing `mode` and `median` functions to complete the task.

- Yes
- No


In [13]:
median_screen_size = laptops_df['Screen'].median()
median_screen_size

15.6

In [14]:
mode_screen_size = laptops_df['Screen'].mode()[0]
mode_screen_size

15.6

In [15]:
# before fillna
laptops_df[laptops_df['Screen'].isnull()]

Laptop Status   Brand  \
624   Acer Extensa 15 EX215-54 Intel Core i5-1135G7/...    New    Acer   
1430  HP ENVY x360 2-in-1 Laptop 15-ew0008np Intel C...    New      HP   
1503  Lenovo IdeaPad Gaming 3 15ACH6 AMD Ryzen 5 560...    New  Lenovo   
1548  Lenovo ThinkPad P15 Gen 2 Intel Core i7-11850H...    New  Lenovo   

         Model            CPU  RAM  Storage Storage type        GPU  Screen  \
624    Extensa  Intel Core i5    8      256          SSD        NaN     NaN   
1430      Envy  Intel Core i7   16      512          SSD   RTX 2050     NaN   
1503   IdeaPad    AMD Ryzen 5   16      512          SSD   RTX 3060     NaN   
1548  ThinkPad  Intel Core i7   16      512          SSD  RTX A2000     NaN   

     Touch  Final Price  
624     No       524.99  
1430   Yes      1863.52  
1503    No      1505.00  
1548    No      2569.00

In [16]:
laptops_df.fillna({'Screen':mode_screen_size}, inplace=True)

In [17]:
# after fillna
laptops_df[laptops_df['Screen'].isnull()]

Empty DataFrame
Columns: [Laptop, Status, Brand, Model, CPU, RAM, Storage, Storage type, GPU, Screen, Touch, Final Price]
Index: []

In [18]:
# after fillna operation the median value of Screen size is:
laptops_df['Screen'].median()

15.6

## Q7. Sum of weights
1. Select all the "Innjoo" laptops from the dataset.
2. Select only columns `RAM`, `Storage`, `Screen`.
3. Get the underlying NumPy array. Let's call it `X`.
4. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
5. Compute the inverse of `XTX`.
6. Create an array `y` with values `[1100, 1300, 800, 900, 1000, 1100]`.
7. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
8. What's the sum of all the elements of the result?
> Note: You just implemented linear regression. We'll talk about it in the next lesson.

- 0.43
- 45.29
- 45.58
- 91.30

In [19]:
innjoo_laptop_df = laptops_df[laptops_df['Brand'] == 'Innjoo']
innjoo_laptop_df

Laptop Status   Brand Model  \
1478  InnJoo Voom Excellence Intel Celeron N4020/8GB...    New  Innjoo  Voom   
1479  InnJoo Voom Excellence Pro Intel Celeron N4020...    New  Innjoo  Voom   
1480  Innjoo Voom Intel Celeron N3350/4GB/64GB eMMC/...    New  Innjoo  Voom   
1481  Innjoo Voom Laptop Max Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1482  Innjoo Voom Laptop Pro Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1483  Innjoo Voom Pro Intel Celeron N3350/6GB/128GB ...    New  Innjoo  Voom   

                CPU  RAM  Storage Storage type  GPU  Screen Touch  Final Price  
1478  Intel Celeron    8      256          SSD  NaN    15.6    No       311.37  
1479  Intel Celeron    8      512          SSD  NaN    15.6    No       392.55  
1480  Intel Celeron    4       64         eMMC  NaN    14.1    No       251.40  
1481  Intel Celeron    6       64         eMMC  NaN    14.1    No       383.61  
1482  Intel Celeron    6      128          SSD  NaN    14.1    No       317.02  
1483  Intel Celeron    6      128         eMMC  NaN    14.1    No       431.38

In [20]:
innjoo_laptop_df.shape

(6, 12)

In [21]:
filtered_innjoo_laptop_df = innjoo_laptop_df[['RAM', 'Storage', 'Screen']]
filtered_innjoo_laptop_df

RAM  Storage  Screen
1478    8      256    15.6
1479    8      512    15.6
1480    4       64    14.1
1481    6       64    14.1
1482    6      128    14.1
1483    6      128    14.1

In [22]:
X = filtered_innjoo_laptop_df.to_numpy()
X

array([[  8. , 256. ,  15.6],
       [  8. , 512. ,  15.6],
       [  4. ,  64. ,  14.1],
       [  6. ,  64. ,  14.1],
       [  6. , 128. ,  14.1],
       [  6. , 128. ,  14.1]])

In [23]:
XTX = np.dot(X.T,X)
XTX

array([[2.52000e+02, 8.32000e+03, 5.59800e+02],
       [8.32000e+03, 3.68640e+05, 1.73952e+04],
       [5.59800e+02, 1.73952e+04, 1.28196e+03]])

In [24]:
inv_XTX = np.linalg.inv(XTX)
inv_XTX

array([[ 2.78025381e-01, -1.51791334e-03, -1.00809855e-01],
       [-1.51791334e-03,  1.58286725e-05,  4.48052175e-04],
       [-1.00809855e-01,  4.48052175e-04,  3.87214888e-02]])

In [25]:
y = np.array([1100, 1300, 800, 900, 1000, 1100])

In [26]:
inv_XTX_XT = np.dot(inv_XTX, X.T)
inv_XTX_XT

array([[ 0.26298349, -0.12560233, -0.40646389,  0.14958687,  0.05244042,
         0.05244042],
       [-0.00110155,  0.00295059,  0.00125892, -0.00177691, -0.00076387,
        -0.00076387],
       [-0.08772226,  0.0269791 ,  0.17140891, -0.0302108 , -0.00153546,
        -0.00153546]])

In [27]:
w = np.dot(inv_XTX_XT, y)
w

array([45.58076606,  0.42783519, 45.29127938])

In [28]:
sum_w = sum(w)
sum_w

91.29988062995588